### PR TITLE
[PGNCCL] Launch kernel on current stream & remove `record_stream` entirely

### LINKED ISSUE #146881: [Feature][c10d] Allow flexible `cudaStreamWaitEvent` in PGNCCL
### 🚀 The feature, motivation and pitch

Today, PGNCCL collectives have the internal NCCL stream wait on the current stream before issuing the NCCL kernel in the NCCL stream. However, waiting on the current stream can be over-synchronization and waiting on an earlier CUDA event could suffice for correctness.

I have a specific use case where it could be useful to pass in a user-specified CUDA event to a PGNCCL collective (in my case, P2P send and recv) and have the collective wait on that instead of recording a new event and waiting on that at collective call time.

The current design mostly assumes that the collective will be called immediately after the relevant kernels needed for that collective; however, when we have multidimensional parallelism, this may not always be the case anymore. There could be several collectives being issued for different PGs, possibly requiring non-collective ops (e.g. device copies) only related to those collectives -- we do not want later collectives to wait on those non-collective ops that have no relevance to them.

One possible API would be to add an `event: Optional[torch.Event]` arg to each c10d collective (though I am not sure how to rationalize this across non-NCCL backends), where if the user passes an event, then the PGNCCL impl should wait on that event instead of doing a stream sync with the current stream.

### Alternatives

We can manually reorder the collectives to avoid over-synchronization, but it may be intrusive to the different parallelism components.

### Additional context

_No response_

cc @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### LINKED ISSUE #147168: [FSDP2] The evil `record_stream` in c10d causes FSDP2 to over-allocate GPU memory
Hey Andrew @awgu,

As a big fan of FSDP2, I find an potential bug 😄

## Demand:
- No inter-stream memory fragmentation (incurred by copy in streams)
- Explicit Prefetch
- CPU runs a head of GPU by a lot

## `_set_unshard_async_op(True)`

To satisfy these demands, FSDP2 has to turn on [`_set_unshard_async_op(True)`](https://github.com/pytorch/pytorch/blob/20a369aa3abb6083600d5b22fcd8ba6e861c3959/torch/distributed/fsdp/_fully_shard/_fully_shard.py#L413) with explicit prefetch `set_modules_to_forward_prefetch` and `set_modules_to_backward_prefetch`.

##  Memory Over-Allocation

Then memory over-allocation happens like this:

Image

with memory traces:

Image

Image

## Root Cause

As known to all, these memory over-allocations are caused by the evil `tensor.record_stream(ncclStream)`. Although FSDP2 tried to avoid this evil originated from FSDP1, such `record_stream` still is [embedded in all c10d collectives](https://github.com/pytorch/pytorch/blob/0acbf8039abccfc17f9c8529d217209db5a7cc85/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L5373) (when `async_op=True`). Therefore, FSDP2 still suffers over-allocation from this evil in c10d.

## Candidate Solution

Not sure how can we avoid the `record_stream` even when `async_op=True`?

IMO, candidate solutions are below:
1. Make `TORCH_NCCL_AVOID_RECORD_STREAMS=True` as an default value, getting rid of the `record_stream` in c10d. (Safety should be good without `record_stream`, as collective with `async_op=True` usually starts from allocation stream and ends at allocation stream, or users indeed know how to [manually sync streams](https://pytorch.org/docs/stable/generated/torch.Tensor.record_stream.html).)

2. Make `TORCH_NCCL_AVOID_RECORD_STREAMS=True`  an advanced option to each collective, such as `dist.all_gather(..., _avoid_record_stream=True)`. This limits the scope of environmental `TORCH_NCCL_AVOID_RECORD_STREAMS` to each specific collective.

3. Use only `dist.all_gather(async_op=False)`  in FSDP2, but [changes the `current_stream`](https://github.com/pytorch/pytorch/blob/20a369aa3abb6083600d5b22fcd8ba6e861c3959/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py#L92) to the `all_gather_stream` such that all gather still allocates/frees in `current_stream` while runs in `all_gather_stream` and overlaps with `current_stream`, just like `async_op=True`.

```python
def get_all_gather_streams(
        self, async_op: bool, training_state: TrainingState
    ) -> tuple[torch.Stream, torch.Stream]:
        if not async_op and training_state in (
            TrainingState.FORWARD,
            TrainingState.PRE_BACKWARD,
        ):
            # Use separate streams for implicit prefetching
            return self.all_gather_copy_in_stream, self.all_gather_stream
        
        # Use separate streams for explicit prefetching!
        current_stream = self.device_handle.current_stream()
        return current_stream, self.all_gather_stream # Change this!
```

How do you prefer? 

(Let us make FSDP great again 😄)

## Code

P.S. the code to reproduce over-allocation:
```python
class MLP(nn.Module):
    def __init__(self, hidden_dim: int, bias: bool = False):
        super().__init__()
        self.fc1 = nn.Linear(hidden_dim, hidden_dim, bias=bias)
        self.gelu = nn.GELU()
        self.fc2 = nn.Linear(hidden_dim, hidden_dim, bias=bias)

    def forward(self, x):
        x = self.fc1(x)
        x = self.gelu(x)
        x = self.fc2(x)
        return x


class MultiMLP(nn.Module):
    def __init__(self, hidden_dim: int, bias: bool = False, layers: int = 4):
        super().__init__()
        self.pre_norm = nn.LayerNorm(hidden_dim, bias=bias)
        self.mlps = nn.ModuleList([MLP(hidden_dim, bias) for _ in range(layers)])
        self.post_norm = nn.LayerNorm(hidden_dim, bias=bias)

    def forward(self, x):
        x = self.pre_norm(x)
        for mlp in self.mlps:
            x = x + mlp(x)
        x = self.post_norm(x)
        return x

class TestMemory(DTensorTestBase):
    @with_comms
    def test_over_allocation(self):
        mesh = init_device_mesh("cuda", (self.world_size,))
        device = torch.device("cuda")
        hidden_dim = 10240
        total_bsz = 16

        # ----- init model --------
        torch.manual_seed(0)
        model = MultiMLP(hidden_dim=hidden_dim).to(device).to(torch.float32)

        # --------  fsdp2 wrap --------
        fully_shard_fn = functools.partial(
            fully_shard,
            mesh=mesh,
            reshard_after_forward=True,
        )

        last_fsdp_module = None
        for module in model.modules():
            if isinstance(module, MLP):
                fully_shard_fn(module)
                if last_fsdp_module is not None:
                    last_fsdp_module.set_modules_to_forward_prefetch([module])
                    module.set_modules_to_backward_prefetch([last_fsdp_module])
                last_fsdp_module = module
        fsdp_model = fully_shard_fn(model)
        fsdp_model._set_unshard_async_op(True)

        optim = torch.optim.Adam(fsdp_model.parameters())

        # ----- init data -----
        torch.manual_seed(self.rank)
        bsz = total_bsz // self.world_size

        # --------  training loop --------
        torch.distributed.barrier()
        torch.cuda.synchronize(self.rank)
        
        train_iter = 4
        for iter in range(train_iter):
            # torch.distributed.barrier()
            # torch.cuda.synchronize(self.rank)

            if self.rank == 0 and iter == train_iter - 1:
                torch.cuda.memory._record_memory_history(max_entries=int(1E6))

            with record_function("## zero grad ##"):
                optim.zero_grad()

            input = torch.randn((bsz, hidden_dim), device="cuda")

            with record_function(f"## forward ##"):
                output = fsdp_model(input)
                loss = output.mean()

            with record_function(f"## backward ##"):
                loss.backward()

            with record_function("## optimizer step ##"):
                optim.step()

            if self.rank == 0 and iter == train_iter - 1:
                timestamp = datetime.now().strftime("%b_%d_%H_%M_%S")
                file_name = f"mem_{timestamp}"
                torch.cuda.memory._dump_snapshot(f"{file_name}.pickle")
                torch.cuda.memory._record_memory_history(enabled=None)

        torch.distributed.barrier()
        torch.cuda.synchronize(self.rank)

```

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @zhaojuanmao @mrshenli @rohan-varma @chauhang

### LINKED ISSUE #147729: [RFE][Distributed][NCCL] A feature request for stream management API in PG NCCL
### A feature request for stream management API in PG NCCL
Asynchronous communication offers the benefit of being overlapped with other CUDA operations, thanks to stream concurrency. However, in the current state of PyTorch, this advantage may be compromised by a potential "read-before-write" issue, as each NCCL process group operates on its own dedicated stream, which may act independently.

Here is an example of a proper behavior:
```python
dist.reduce_scatter_tensor(buffer, data, group=local_pg)
dist.all_reduce(buffer, group=cross_pg)
dist.all_gather_into_tensor(data, buffer, group=local_pg)
```
Image
Where the collective ops are executed in the right order.

However, once these ops are requested as async:
```python
dist.reduce_scatter_tensor(buffer, data, group=local_pg, async_op=True)
dist.all_reduce(buffer, group=cross_pg, async_op=True)
dist.all_gather_into_tensor(data, buffer, group=local_pg, async_op=True)
```
Image
The collective ops are no longer in proper order since they run on different streams that are asynchronous wrt each other. **This behavior is certainly not what users would expect and can be considered as a bug.**

One can mitigate this with `handle.wait()` call, but the main goal is to run these comms parallel to the compute ops (matmuls etc.). We would like to propose one of two ways of setting streams in process groups:

- Pass a user defined stream as PG-option in new_group:
```python
nccl_options = dist.ProcessGroupNCCL.Options(stream=user_stream)
pg = dist.new_group(backend="nccl", pg_options=nccl_options)
```

- Set a user stream at runtime:
```python
pg1.set_stream(user_stream)
pg2.set_stream(user_stream)
dist.reduce_scatter_tensor(buffer, data, group=pg1, async_op=True)
dist.all_reduce(buffer, group=pg2, async_op=True)
```

### Motivation
It was first discovered in an attempt to do multi data center training. In the above example
```python
dist.reduce_scatter_tensor(buffer, data, group=local_pg)
dist.all_reduce(buffer, group=cross_pg)
dist.all_gather_into_tensor(data, buffer, group=local_pg)
```
The `cross_pg group` has “netName” set to “Socket” in ProcessGroupOptionsNCCL to tell NCCL to use a frontend TCP network which connects two compute clusters together, instead of IB or RoCE, that are set for `local_pg`. Without more control over the communication stream, the inter DC communication must be exposed and not overlapped with any compute.

There are many other cases that could potentially benefit from fine control of compute stream as well.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @ptrblck @msaroufim @eqy @skyw @alpha0422

### PR DESCRIPTION
**Pull request by kwen2501**
This PR has multiple changes to `ProcessGroupNCCL` (which unfortunately have to be atomic):
1. When async_op=False, we directly launch the collective on "current" stream instead of a trampoline stream and join back. 
- Resolves #147729
- Resolves #146881
- Also saves an event sync and one pybind during the unnecessary `work.wait()` called by distributed_c10d.py.
2. Entirely remove `record_stream` and use CPU-side stashing for managing tensor lifetime against recycling.
- Resolves #147168
3. Remove tensor life management when async_op=False; only use it when async_op=True.
4. To guard against user not calling `work.wait()`, we ask watchdog to unstash tensors after detecting completion of collectives, to prevent us from holding reference to tensors forever. This is a safety net, rather than a service guarantee, see discussion [here](https://github.com/pytorch/pytorch/issues/147168#issuecomment-2660142460).
5. Profile in async_op=False mode would look different -- collective kernels would show up in the same line and compute kernels.

Joint work with @cenzhaometa who wants to remove the event sync overhead.

Cc: @ngimel @awgu @Aidyn-A @skyw @wconstab @leonardo0lyj

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -3643,13 +3636,8 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
...
   // and the present call has no way to know it's not an isend.
-  // Therefore, we warn and fall back to the typical recordStream logic:
-  if (avoidRecordStreams_) {
-    TORCH_WARN_ONCE(
-        "TORCH_NCCL_AVOID_RECORD_STREAMS=1 has no effect for point-to-point "
-        "collectives.");
-  }
-
+  // Therefore, we warn and fall back to the typical recordStream logic.
+  // TODO( kwen2501 ): revisit this when we have a better solution.
```
Comment (eqy) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:3640]: A few options here:
1. Change the requirement to be that the user has to call `.wait()` even on isend calls.
2. Add some logic to the destructor that does the `synchronizeStream` or something equivalent to block the destruction on the p2p op

**Comment by kwen2501** [comment]
Moving over to #148590 as internal user needs `ghimport` which supports `ghstack` PRs only.
